libs/wingc/src/closure_transform.rs

```diff
@@ -0,0 +1,306 @@
+				};
+
+				let class_type_annotation = TypeAnnotation {
+					kind: TypeAnnotationKind::UserDefined(UserDefinedType {
+						root: new_class_name.clone(),
+						fields: vec![],
+					}),
+					span: expr.span.clone(),
+				};
+
+				let class_fields: Vec<ClassField> = vec![];
+				let class_init_params: Vec<FunctionParameter> = vec![];
+
+				let new_func_def = if self.inside_scope_with_this {
+					// If we are inside a class, we transform inflight closures with an extra
+					// `let __parent_this = this;` statement before the class definition, and replace all references
+					// of `this` with `__parent_this` so that they can access the parent class's fields.
+					let mut this_transform = RenameIdentifierTransformer {
+						from: "this",
+						to: PARENT_THIS_NAME,
+					};
+					this_transform.fold_function_definition(func_def)
+				} else {
+					func_def
+				};
+
+				// class_def :=
+				// ```
+				// class <new_class_name> {
+				//   init(<class_init_params>) {<class_init_body>}
+				//   inflight handle() {
+				//     <transformed_def>
+				//   }
+				// }
+				// ```
+				let class_def = Stmt {
+					kind: StmtKind::Class(Class {
+						name: new_class_name.clone(),
+						is_resource: true,
+						initializer: Initializer {
+							signature: FunctionSignature {
+								parameters: class_init_params,
+								return_type: Some(Box::new(class_type_annotation.clone())),
+								phase: Phase::Preflight,
+							},
+							statements: Scope::new(vec![], expr.span.clone()),
+							span: expr.span.clone(),
+						},
+						fields: class_fields,
+						implements: vec![],
+						parent: None,
+						methods: vec![(handle_name.clone(), new_func_def)],
+						inflight_initializer: None,
+					}),
+					idx: self.nearest_stmt_idx,
+					span: expr.span.clone(),
+				};
+
+				// new_class_instance :=
+				// ```
+				// new <new_class_name>();
+				// ```
+				let new_class_instance = Expr::new(
+					ExprKind::New {
+						class: class_type_annotation,
+						arg_list: ArgList {
+							named_args: IndexMap::new(),
+							pos_args: vec![],
+						},
+						obj_id: None,
+						obj_scope: None,
+					},
+					expr.span.clone(),
+				);
+
+				self.class_statements.push(class_def);
+
+				new_class_instance
+			}
+			_ => fold::fold_expr(self, expr),
+		}
+	}
+}
+
+/// Rename all occurrences of an identifier to a new name.
+struct RenameIdentifierTransformer<'a> {
+	from: &'a str,
+	to: &'a str,
+}
+
+impl<'a> Fold for RenameIdentifierTransformer<'a> {
+	fn fold_reference(&mut self, node: Reference) -> Reference {
+		match node {
+			Reference::Identifier(ident) => {
+				if ident.name == self.from {
+					Reference::Identifier(Symbol {
+						name: self.to.to_string(),
+						span: ident.span,
+					})
+				} else {
+					Reference::Identifier(ident)
+				}
+			}
+			Reference::InstanceMember { .. } => fold::fold_reference(self, node),
+			Reference::TypeMember { .. } => fold::fold_reference(self, node),
```

Comment:
```suggestion
			_ => fold::fold_reference(self, node),
```

Reply:
I would rather make the cases explicit just so that if we add or remove choices to this enum in the future, someone is required to update this transformer